### PR TITLE
Added tests for singleton subsystem

### PR DESCRIPTION
**Pull request by honza-kasik**

### CODE REVIEW

**Comment by honza-kasik** [comment]
~~Ready to be merged~~
~~Found bug, WIP~~
Ready to be merged, finally...

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            CACHE_CONTAINER = "cache-container",
+            SINGLETON_POLICY = "singleton-policy",
+            ELECTION_POLICY = "election-policy",
+            NAME_PREFERENCES = "name-preferences",
+            SOCKET_BINDING_PREFERENCES = "socket-binding-preferences",
+            SIMPLE = "simple";
+
+    private static final Address
+            SINGLETON_SUBSYSTEM_ADDRESS = ConfigUtils.isDomain() ? Address.of("profile", "full-ha").and("subsystem", "singleton") : Address.subsystem("singleton"),
+            INFINISPAN_SUBSYSTEM_ADDRESS = ConfigUtils.isDomain() ? Address.of("profile", "full-ha").and("subsystem", "infinispan") : Address.subsystem("infinispan"),
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: The check for doamin isn't needed, the prefix for subsystem in domain is done based on management client configuration.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        page.navigate();
+    }
+
+    @BeforeClass
+    public static void beforeClass() throws IOException, CommandFailedException {
+        socketBindingResources = new SocketBindingResources(client, 3, "election-policy_");
+        SingletonSubsystemOperations.withOperations(operations).prepareCacheContainer(CACHE_CONTAINER_RESOURCE_ADDRESS);
+        operations.add(SINGLETON_POLICY_ADDRESS, Values.of(CACHE_CONTAINER, CACHE_CONTAINER_RESOURCE)).assertSuccess();
+        operations.add(SINGLETON_POLICY_EDIT_ADDRESS, Values.of(CACHE_CONTAINER, CACHE_CONTAINER_RESOURCE)).assertSuccess();
+        operations.add(SINGLETON_POLICY_EDIT_ADDRESS.and(ELECTION_POLICY, "simple")).assertSuccess();
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: I suggest using batch operation, which takes care of cleaning after itself if not all of the substeps succeed.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                .navigateToTreeItem().clickLabel();
+
+        new ConfigChecker.Builder(client, SINGLETON_POLICY_EDIT_ADDRESS.and(ELECTION_POLICY, SIMPLE))
+                .configFragment(page.getConfigFragment())
+                .editAndSave(ConfigChecker.InputType.TEXT, SOCKET_BINDING_PREFERENCES, socketBindingResources.getSocketBindingNamesSeparatedByNewline())
+                .verifyFormSaved()
+                .verifyAttribute(SOCKET_BINDING_PREFERENCES, socketBindingResources.getSocketBindingsAsModelNodeList());
+    }
+
+    private static final class SocketBindingResources implements AutoCloseable {
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Why AutoCloseable when you are not using it this way, I suggest just providing clean method, which is in this case more representative. AutoCloseable is mainly meant for resources which have allocated some resources needing to be unallocated, such as streams, connections. Not as cleanup method of what was done. You can create interface for the purpose of cleaning if you like.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        operations.add(SINGLETON_POLICY_EDIT_ADDRESS.and(ELECTION_POLICY, "simple")).assertSuccess();
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
+        try {
+            operations.removeIfExists(SINGLETON_POLICY_ADDRESS);
+            operations.removeIfExists(SINGLETON_POLICY_EDIT_ADDRESS);
+            SingletonSubsystemOperations.withOperations(operations).removeCacheContainer(CACHE_CONTAINER_RESOURCE_ADDRESS);
+            socketBindingResources.close();
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: should be probably in some finally block, the removeCacheContainer may throw exception

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        operations.add(SINGLETON_POLICY_EDIT_ADDRESS, Values.of(CACHE_CONTAINER, CACHE_CONTAINER_RESOURCE)).assertSuccess();
+        operations.add(SINGLETON_POLICY_EDIT_ADDRESS.and(ELECTION_POLICY, "simple")).assertSuccess();
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
+        try {
+            operations.removeIfExists(SINGLETON_POLICY_ADDRESS);
+            operations.removeIfExists(SINGLETON_POLICY_EDIT_ADDRESS);
+            SingletonSubsystemOperations.withOperations(operations).removeCacheContainer(CACHE_CONTAINER_RESOURCE_ADDRESS);
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Use commonly used pattern for creating custom operations class => instantiate via OnlineManagementClient client and not operations. Also don't use Factory pattern for this purpose, it should stay consistent with the other Operations classes.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+            }
+            return socketBindingNames;
+        }
+
+
+        public String getSocketBindingNamesSeparatedByNewline() {
+            return socketBindingsNames.stream().collect(Collectors.joining("\n"));
+        }
+
+        public ModelNode getSocketBindingsAsModelNodeList() {
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java:220]: Javadoc

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                operations.add(remoteSocketBindingAddress,
+                        Values.of("host", "el-policy-host_" + RandomStringUtils.randomAlphanumeric(5))
+                            .and("port", AvailablePortFinder.getNextAvailableNonPrivilegedPort())).assertSuccess();
+                socketBindingNames.add(name);
+            }
+            return socketBindingNames;
+        }
+
+
+        public String getSocketBindingNamesSeparatedByNewline() {
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java:212]: Javadoc please

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+            this.socketBindingGroupAddress = Address.of("socket-binding-group", contextSocketBindingGroup);
+
+            this.socketBindingsNames = createSocketBindings();
+        }
+
+        private String generateName() {
+            return prefix + RandomStringUtils.randomAlphanumeric(5);
+        }
+
+        private List<String> createSocketBindings() throws CommandFailedException, IOException {
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Why throws `CommandFailedException`?

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java (hunk 3 above)
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Javadoc please

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        }
+
+        public ModelNode getSocketBindingsAsModelNodeList() {
+            final ModelNode node = new ModelNode();
+            socketBindingsNames.forEach(node::add);
+            return node;
+        }
+
+        @Override
+        public void close() throws Exception {
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Javadoc please

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/SingletonPolicyTestCase.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+            CACHE_CONTAINER_RESOURCE = "cache-container-singleton_" + RandomStringUtils.randomAlphanumeric(5),
+            CACHE_CONTAINER_RESOURCE_EDIT = "cache-container-singleton-edit_" + RandomStringUtils.randomAlphanumeric(5),
+            CACHE_CONTAINER = "cache-container",
+            CACHE = "cache",
+            QUORUM = "quorum",
+            NAME = "name";
+
+    private static final Address
+            SINGLETON_SUBSYSTEM_ADDRESS = ConfigUtils.isDomain() ? Address.of("profile", "full-ha").and("subsystem", "singleton") : Address.subsystem("singleton"),
+            INFINISPAN_SUBSYSTEM_ADDRESS = ConfigUtils.isDomain() ? Address.of("profile", "full-ha").and("subsystem", "infinispan") : Address.subsystem("infinispan"),
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/SingletonPolicyTestCase.java]: As above, no need the special handling for domain.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/SingletonSubsystemOperations.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    private Operations operations;
+
+    /**
+     * Create new instance
+     * @param operations used instance of operations
+     * @return new instance
+     */
+    public static SingletonSubsystemOperations withOperations(Operations operations) {
+        return new SingletonSubsystemOperations(operations);
+    }
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/SingletonSubsystemOperations.java]: Please don't use this pattern for Operations classes, it makes inconsistent usage for them and makes it rather confusing than helping.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/SingletonSubsystemOperations.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+    private SingletonSubsystemOperations(Operations operations) {
+        this.operations = operations;
+    }
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/SingletonSubsystemOperations.java:24]: Make constructor public and the attribute it gets should be `OnlineManagementClient`

**Review by rhatlapa — CHANGES_REQUESTED** [review]
There is no need for subsystem addresses to be special counted for domain, this is automatically handled by creaper based on client config.

Also don't change pattern for usage of different Creaper Operations classes, we should stay consistent. It will be nicer from dev UX.

**Comment by honza-kasik** [comment]
Thanks! Worked in suggestions, ready to be merged.

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 13]
```diff
@@ -54,16 +54,18 @@
...
-            SINGLETON_SUBSYSTEM_ADDRESS = ConfigUtils.isDomain() ? Address.of("profile", "full-ha").and("subsystem", "singleton") : Address.subsystem("singleton"),
-            INFINISPAN_SUBSYSTEM_ADDRESS = ConfigUtils.isDomain() ? Address.of("profile", "full-ha").and("subsystem", "infinispan") : Address.subsystem("infinispan"),
+            SINGLETON_SUBSYSTEM_ADDRESS = Address.subsystem("singleton"),
+            INFINISPAN_SUBSYSTEM_ADDRESS = Address.subsystem("infinispan"),
             SINGLETON_POLICY_EDIT_ADDRESS = SINGLETON_SUBSYSTEM_ADDRESS.and("singleton-policy", SINGLETON_POLICY_EDIT_RESOURCE),
             SINGLETON_POLICY_ADDRESS = SINGLETON_SUBSYSTEM_ADDRESS.and("singleton-policy", SINGLETON_POLICY_RESOURCE),
             CACHE_CONTAINER_RESOURCE_ADDRESS = INFINISPAN_SUBSYSTEM_ADDRESS.and("cache-container", CACHE_CONTAINER_RESOURCE);
 
-    private static final OnlineManagementClient client = ManagementClientProvider.createOnlineManagementClient();
+    private static final OnlineManagementClient client = ManagementClientProvider.withProfile("full-ha");
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Where did disapear client for standalone?

**Review by rhatlapa — CHANGES_REQUESTED** [review]

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java (hunk 13 above)
Comment (honza-kasik) [reply basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java]: Thanks! Fixed!

File: basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+            return node;
+        }
+
+        /**
+         * Removes all created socket bindings
+         * @throws IOException
+         */
+        public void clean() throws IOException {
+            for (String socketBinding : socketBindingsNames) {
+                operations.remove(this.socketBindingGroupAddress.and(SOCKET_BINDING_TYPE, socketBinding)).assertSuccess();
```
Comment (rhatlapa) [inline basic/src/test/java/org/jboss/hal/testsuite/test/configuration/singleton/ElectionPolicyTestCase.java:232]: removeIfExists might be more reasonable in this case, but ok.

**Review by rhatlapa — APPROVED** [review]

**Comment by rhatlapa** [comment]
LGTM, please squash and I shall merge.

**Comment by honza-kasik** [comment]
Squashed and ready t be merged!

**Comment by rhatlapa** [comment]
Thanks, merging.

**Pull request merged**